File: src/14장/14-3-3.py
def back_Traking(index,sum):
    global minAns
    global maxAns
    if index==N-1:
        if minAns > sum:minAns = sum
        if maxAns < sum:maxAns = sum
        return 

    for i in range(4):
        temp = sum
        if operator[i]==0:continue
        if i==0:sum+=numArr[index+1]
        elif i==1:sum-=numArr[index+1]
        elif i==2:sum*=numArr[index+1]
        else:
            if sum<0:sum = abs(sum)//numArr[index+1]*-1
            else:sum //=numArr[index+1]

        operator[i] -= 1
        back_Traking(index+1,sum)
        operator[i] += 1
        sum = temp

N = int(input())
numArr = list(map(int,input().split()))
operator = list(map(int,input().split()))
minAns = float('Inf')
maxAns = float('-Inf')

back_Traking(0,numArr[0])
print(maxAns)
print(minAns)

File: src/14장/14-3-3.java
import java.util.*;
import java.io.*;

public class Main {
	public static int maxAns = Integer.MIN_VALUE;	// 최댓값 
	public static int minAns = Integer.MAX_VALUE;	// 최솟값 
	public static int[] operator = new int[4];	// 연산자 개수 
	public static int[] numArr;					// 숫자 
	public static int N;						// 숫자 개수 
 
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		numArr = new int[N];
		for (int i = 0; i < N; i++) {
			numArr[i] = sc.nextInt();
		}
		for (int i = 0; i < 4; i++) {
			operator[i] = sc.nextInt();
		}
		back_Traking(numArr[0], 1);
		System.out.println(maxAns);
		System.out.println(minAns);
	}
 
	public static void back_Traking(int num, int idx) {
		if (idx == N) {
			maxAns = Math.max(maxAns, num);
			minAns = Math.min(minAns, num);
			return;
		}
		for (int i = 0; i < 4; i++) {
			if (operator[i] > 0) {
				operator[i]--;
				if(i==0)back_Traking(num + numArr[idx], idx + 1);
				else if(i==1)back_Traking(num - numArr[idx], idx + 1);
				else if(i==2)back_Traking(num * numArr[idx], idx + 1);
				else if(i==3)back_Traking(num / numArr[idx], idx + 1);
				operator[i]++;
			}
		}
	}
}